### ISSUE TITLE
Please clarify: 200 OK response when updating relationships

### ISSUE THREAD

**Issue by maurei** [json-api/json-api#1524]
Please clarify the motive behind the following [bit of the spec](https://jsonapi.org/format/#crud-updating-relationship-responses-200)

> 200 OK
> If a server accepts an update but also changes the targeted relationship(s) in other ways than those specified by the request, it MUST return a 200 OK response. The response document MUST include a representation of the updated relationship(s).

Specifically: what are some use-cases for this?

We have been working hard on improving [JSON:API .NET Core](www.jsonapi.net) and making it fully spec compliant. We understand from a technical perspective how such side-effects could arise, but we are lacking knowledge of any clear use-cases for it, and because introducing it could come with a significant performance cost we have decided to hold off implementing this element of the specification.

We came up with some examples to illustrate this.

1. Given a resource `a` of type `A` with relationships `has-one B` and `has-one C`, we can imagine how setting `B=b1` and `C=c1`  would trigger some database procedure that e.g. would effectively yield `B=b1` and `c=null`.
However, our concern with this is that
	- we don't know of any relevant real-life examples for this, and
	- the performance implications of checking for such effects could be disastrous in the case of a `has-many` relationship when the size of the collection is very large. 
2. The next case is similar to the previous, but this one *does* come with a real-life use-case. Consider a resource type `Article` with a **required** relationship to `User` (`has one Author`). Consider an API user that creates an article (`POST /articles`) without providing a reference for `Author`, and let's assume that in the context of this API it is guaranteed that user who creates the article is always its author. The server could then take care of setting this required relationship by deducing the author from the request context, freeing the API user from the responsibility of passing a (trivial) reference to himself. The problem with this is however:
	- the spec specifically states that side-effects should only be checked for *targeted relationship(s)*, i.e. for the current scenario a 201 No Content would suffice. This makes sense, because performance implications of having to check for every non-targeted relationship of a resource would quickly become unmanageable for any non-trivial model. 

Some explanation would be greatly appreciated!

Thank you

**Comment by gabesullice** [comment]
Caveat: I'm speaking as an implementer here, not as a spec editor because I don't know exactly what the original intent was.

---
Two thoughts:

1. The quoted rule only applies to `PATCH` requests to relationship endpoints, but your first example is updating two separate relationships. IOW, your example is updating via the URL `/resources/a` not `/resources/a/relationships/B` and the rule only applies to the latter case. I think your second example is making a similar mistake since you're trying to apply the rule to `POST /articles`.

2. For Drupal's JSON:API, I interpreted this to mean that if the relationship data in the request document is different than the relationship data after the server is done processing, then the client should be sent than data so that it can update its memory, if it has it.

To do this, all I do is diff the request `data` against the response `data`. If there is no difference I send a `204 No Content`  because the client can assume its present representation of the relationship is up-to-date. However, if there _is_ a difference, I send a `200` so that the client knows to update the resource object in memory.

In other words, my code is written as if I'm going to send a response document and at the last moment I check to see if I should send it or not.

The same pattern would work for your second example: diff the `POST`ed resource object against the result resource object and send decide whether a `201` or `204` is appropriate based on whether the client's representation is outdated or not.

**Comment by bart-degreed** [comment]
Hi @gabesullice, thanks for sharing your interpretation.

@maurei has some days off, so I'm jumping in. We believe that change detection also applies to PATCH on resource endpoints (PATCH /articles/1) because of [this](https://jsonapi.org/format/1.1/#crud-updating-responses-204):
> If an update is successful and the server doesn’t update any fields besides those provided, the server MUST return either a 200 OK status code and response document (as described above) or a 204 No Content status code with no response document.

This refers to _fields_ (so attributes and relationships). But does not scope that down to _targeted_ fields.

If an unsent last-modified attribute changes as a side-effect during partial patch (so non-targeted), we need to detect that. So we fetch all attributes upfront, then save, re-fetch and diff. This way we properly detect changes caused by database triggers. Likewise we tend to think the same somehow applies for relationships. So if a PATCH on resource including relationship A was targeted, causing a change to untargeted relationship B, we need to return 200. But the strange things is that the response will not actually include relationship B, unless it was explicitly asked for using `?include=`. This would mean we need to fetch all relationships with their data upfront, and afterwards, then diff, which is very expensive. So we question how that is helpful for the caller, given we're not sending out the actual changes.

This may be a misinterpretation of the spec from our side, or an omission in the spec. Unless there's an actual use case which we're not aware of, where all of this makes perfect sense.

If, instead all we need to do is diff the request body against the response body before sending out the response to a relationship endpoint, we'd be sending `data { type = "a", id = '1' }`, unless that has changed unexpectedly. We cannot think of a scenario where that would happen, ie the user sent that, but for some reason we're returning `data { type = "b", id = '2' }` instead. What can this "some reason" possibly be?

Hope this clarifies our concerns.

**Comment by lode** [comment]
(Disclaimer: I'm also an implementer.)

For updating resources I also interpreted the spec as that the response needs to be sent if a for example `lastModified` field was updated. I didn't think about the scenario where a relationship gets updated due to this. But from a practical point of view I would ignore that. So a `PATCH /article/1` would not return any relationships even if they change, and a request to `PATCH /article/1?include=author` would include the author anyway, no matter if it changes.

For updating relationships I think the spec is a bit incorrect.

Next to:

> If a server accepts an update but also changes the targeted relationship(s) in other ways than those specified by the request, it MUST return a 200 OK response. The response document MUST include a representation of the updated relationship(s).

It also states at the `204 No Content` response:

> Note: This is the appropriate response to a POST request sent to a URL from a to-many relationship link when that relationship already exists. It is also the appropriate response to a DELETE request sent to a URL from a to-many relationship link when that relationship does not exist.

I don't understand this note. I imagine this is about to-**one** relationships, not to-**many**. Because for a to-many (as POSTing is adding to, not replacing) there's other relationships which are unknown. So I think the note should be:

> Note: This is the appropriate response to a POST request sent to a URL from a to-one relationship link when that relationship already exists. It is also the appropriate response to a DELETE request sent to a URL from a to-one relationship link when that relationship does not exist.

And possible something should be mentioned about to-many as well, like:

> Note: This is the appropriate response to a POST request sent to a URL from a to-many relationship link when that relationship is the only relationship (already existing or existing after processing the request). It is also the appropriate response to a DELETE request sent to a URL from a to-many relationship link when that relationship does not exist or was the last relationship.

Or would it be allowed to reply with a 204 on to-many relationship requests in all cases when they are successful? That would for sure safe some server/transfer resources and probably be meaningless to the client anyway. Unless the client wants to update what other clients did in the mean time, but that feels like a minor feature compared to making it required in the spec.

(Btw: does anyone know different words for 'relationship'? Like we name `authors` on `Article` a relationship, and we name `{'data': [{'type': 'Human', id: '42'}]}` a relationship.)

**Comment by lode** [comment]
> I don't understand this note. I imagine this is about to-**one** relationships, not to-**many**. Because for a to-many (as POSTing is adding to, not replacing) there's other relationships which are unknown. So I think the note should be:
>
>> Note: This is the appropriate response to a POST request sent to a URL from a to-one relationship link when that relationship already exists. It is also the appropriate response to a DELETE request sent to a URL from a to-one relationship link when that relationship does not exist.

I'm reconsidering this. For to-one relationships it seems one can always return a `204` if it was a success. Whether it changed something or was a race condition doesn't matter.

Otherwise it is still a confusing text for to-many relationships.

**Comment by jelhan** [comment]
I agree that the [responses sections for updating relationships](https://jsonapi.org/format/1.1/#crud-updating-relationship-responses) is confusing and maybe slightly wrong.

>  204 No Content
>
> A server MUST return a 204 No Content status code if an update is successful and the representation of the resource in the request matches the result.
>
> https://jsonapi.org/format/1.1/#crud-updating-relationship-responses

The request is _not_ about a resource. It is about a relationship between two resources.

Further more the specification currently does not support responding with the resource for a request to update a relationship. This limitation has been raised by @cmeeren in #1492. I don't see what a server should do, if updating a relationship causes the resource to change.

On example for this situation would be adding to or a deleting from a `reviews` relationship of a `product` resource, which changes an `averageReviewStars` attribute of that resource.

>  200 OK
> 
> If a server accepts an update but also changes the targeted relationship(s) in other ways than those specified by the request, it MUST return a 200 OK response. The response document MUST include a representation of the updated relationship(s).
>
> https://jsonapi.org/format/1.1/#crud-updating-relationship-responses

What is a "relationship of the updated relationship(s)"? Should the server response with resource identifier objects of all resources, which are associated with that relationship after processing the update? Or should it only response with the affected items in the relationship?

It is especially confusing that the specification talks about "relationship(s)". Using a relationship link only one single relationship could be updated per request. For to-many relationships such a request could add or delete multiple resources from that relationship.

I think some rewriting of that chapter is needed to better express the intent.

**Comment by cmeeren** [comment]
I haven't thoroughly read all comments in this issue, but based on the OP it seems that this may be a duplicate of #1494?

**Comment by jelhan** [comment]
This has been addressed by #1638.

**Comment by bkoelman** [comment]
Please reopen this, as the concern [has not been addressed](https://github.com/json-api/json-api/pull/1638/files/54d2a4d013e6d2da7670d4efac1d5b1ce3e4ad7d#r907492147).

**Comment by jelhan** [comment]
> Please reopen this, as the concern [has not been addressed](https://github.com/json-api/json-api/pull/1638/files/54d2a4d013e6d2da7670d4efac1d5b1ce3e4ad7d#r907492147).

Do I understand correctly that you have performance concerns with current processing rules enforced by the specification for update requests which cause side-effects?

If so: For which of the following request types do you have performance concerns caused by informing clients about side-effects?

1. Updating a resource
2. Updating a to-one relationship
3. Replacing a to-many relationships (`PATCH` to a relationship URL)
4. Adding a resource to or removing a resource from a to-many relationships (`POST` or `DELETE` to a relationship URL)

**Comment by bkoelman** [comment]
No. I'm asking to reopen, because the spec is still ambiguous for the 200 response of relationship endpoints.

For post/patch resource endpoints, a 200 response is defined as:

> If a server accepts an update but also changes the targeted resource in ways other than those specified by the request (for example, updating the `updatedAt` attribute or a computed `sha`), it MUST return a 200 OK response and a document that contains the updated resource as primary data.

This is pretty clear: If a post/patch resource endpoint includes relationship changes, their "side effects" are _not_ taken into account. With "side effects" I mean: any additional changes that come as a surprise to the client because they were not requested (possibly resulting from tailor-made business rules, database triggers, cascading updates/deletes, stored procedures, automatic conflict resolution in optimistic concurrency, etc).

For post/patch/delete relationship endpoints, a 200 response is defined as:

> If a server accepts an update but also changes the targeted relationship in other ways than those specified by the request, it MUST return a 200 OK response and a document that includes the updated relationship data as its primary data.

This is open to interpretation. "changing the targeted relationship in other ways..." could mean either:
1. Any side-effect changes to the identity or attributes of all related resource(s)
2. Any side-effect changes to the identity or attributes of the (subset of) related resource(s) in the request body
3. Any side-effect attachment/detachment in the (set of) related resource(s)
4. Any side-effect attachment/detachment in the (subset of) related resource(s) in the request body

1 and 2 are unlikely, because the spec mentions "relationship" instead of "related resources", and there's no syntax to return the attribute changes in related resources. 3 is the most likely, but only doable for to-one or very small to-many relationships (there's no pagination, and detection is expensive because it requires fetching all type/id pairs). 4 could be done, but I don't see how that's useful to a client: it won't enable cache refresh, because it doesn't include side-effect attachment/detachment of related resources that weren't in the request. For example: if a user removes one phone number from a contact, it won't observe that a concurrent user added two new phone numbers to that same contact.

Likewise, returning "a document that includes the updated relationship data" is open to interpretation as well. It could mean:
1. The type/id pair(s) of the (set of) related resource(s), or data:null
2. Only the (subset of) type/id pair(s) that changed resulting from side effects
3. Only the subset of type/id pairs that were in the request body
4. Only the subset of type/id pairs that were in the request body and that changed resulting from side effects

1 is the most likely, but only doable for to-one or very small to-many relationships. 2 would be doable, but obtaining that information is expensive because it requires fetching all type/id pairs, while I can't think of a practical use case, and there's no syntax to express something was detached as a side effect. 2, 3 and 4 don't enable cache refresh for reasons explained above.

So far, the motivation for why I believe this issue should be reopened. I don't understand _how to interpret_ the spec.

Now the answer to your question, from the perspective of implementing the server-side of JSON:API: My concern only applies to post/patch/delete relationship endpoints. This is the part where I'm saying _I don't like_ what the spec requires (assuming my guesswork from above holds).

1. Returning the complete unpaginated set of type/id pairs for a to-many relationship is a no-go in general, due to the potentially massive response size. This goes against the strength of JSON:API, which enables clients to control network bandwidth by getting _only_ what they asked for (sparse fieldsets, filters, includes, pagination).
2. Even if the returned set is small, there's a tradeoff to make between performance and usability. It's expensive to fetch all type/id pairs (potentially thousands or millions), only to conclude there are no side effects. So unless there's a compelling use case where it pays off, we're not going to do that, as it benefits nobody.
3. We could implement this for to-one relationships, but it's unlikely we'll ever do that unless there's a need for it. It still requires an extra query. That extra query is an acceptable tradeoff for post/patch resource requests, but it doesn't make sense for relationships.

Therefore, I'd like the MUST in 200 relationship responses to be changed to MAY, so that our current behavior is not a violation. In other words: allow to return ~Not Modified~ No Content, even if there are potential side effects. Right now the spec guarantees clients there are no side effects when ~Not Modified~ No Content is returned, which I find unrealistic to require that guarantee from servers.

**Comment by jelhan** [comment]
> If a post/patch resource endpoint includes relationship changes, their "side effects" are _not_ taken into account. With "side effects" I mean: any additional changes that come as a surprise to the client because they were not requested

I read the specification differently. Adding a resource to or removing a resource from a relationship is a change of a resource. If that changes affects the "targeted resource" and was not "specified by the request", the server must response with the updated resource. It doesn't matter if the side-effects affects an attribute or a relationship of the targeted resource. At least that's how I understand the specification.

Maybe that different understanding is the root cause for the confusion?

**Comment by bkoelman** [comment]
I agree with your interpretation. In that case, it has the same problem. For example, adding an audit log entry as a side effect of resource attribute update requires to return the complete list of audit entry type/id pairs, which is totally impractical.

**Comment by jelhan** [comment]
Great. Let's get back to your second question on the expected primary data when responding to an update request with `200 Ok`.

For both cases, updating a resource as well as updating a relationship using a relationship URL, the response should be the same as if client done a `GET` request to the same URL immediately afterwards. The primary intent is informing the client about resulting state preventing the need of an additional `GET` request.

I hear your performance concerns. But aren't those affecting the `GET` requests as well? How do you address them on fetching? Why can't you use the same solution for responses to mutation requests?

**Comment by cmeeren** [comment]
@bkoelman 

> It's expensive to fetch all type/id pairs (potentially thousands or millions)

What is the use of such large relationships, given that no client would likely ever want the complete set? Could it be modeled another way, to avoid such large relationships?

For example, in our APIs, we have audit logs (we call this resource `event`) for most changes to all resources. But we don't have a relationship `events` on each resource, nor do we have a relationship from the `event` resource to the target resource. Instead, the `event` resource simply has a `resourceId` attribute, and if a client wishes to fetch audit logs for a resource, they call

```http
GET /events?filter[resourceId]=12345&page[limit]=100&page[offset]=0
```

so they have full control of pagination.

We have over 30 APIs and probably at least a hundred, if not several hundred, to-many relationships across the APIs, all using my JSON:API framework [Felicity](https://github.com/cmeeren/Felicity). Felicity always responds with all resource linkage for POST/PATCH/DELETE to to-many relationship `self` endpoints. This issue is not a problem for us, because we simply don't have to-many relationships that have thousands or millions of items. (If we had, an unwary client could add `include=expensiveRelationship` to the request, and then suddenly we'd have huge performance issues, potential out-of-memory exceptions, etc.)

I could also mention that the only way to create resources in Felicity is POST to a "base" resource collection endpoint. It does not support POST to relationship `related` endpoints. All resource creation requests happens using `POST` to e.g. `/orderLines`, with a back-reference `order` relationship, not by `POST /orders/{id}/orderLines`. This works great.

**Comment by bkoelman** [comment]
> I hear your performance concerns. But aren't those affecting the GET requests as well? How do you address them on fetching? Why can't you use the same solution for responses to mutation requests?

> What is the use of such large relationships, given that no client would likely ever want the complete set? Could it be modeled another way, to avoid such large relationships?

Perhaps it wasn't clear that JsonApiDotNetCore is a framework that enables developers to model their own resources and fields, along with database mappings. The framework exposes them through JSON:API and generates SQL to fetch/update the underlying data, removing the need for writing boilerplate code. So _we_ don't model anything, that's up to our users. However, we aim to provide a scalable experience out of the box. That's why pagination is turned on by default. So a GET request would be paginated, unless a client sends `?page[size]=0` and the framework is explicitly configured to allow that. Furthermore, a GET request returns relationships _without_ their data, unless `?include=...` is used (so opt-in). Related resources that are included get paginated as well. We have pagination syntax to allow fetching the first/last/prev/next page of related resources. Traversing the paginated related resources can be done efficiently by deselecting all top-level attributes on subsequent requests.
For example: `GET /blogs/1?include=posts` returns the blog attributes, along with the first 10 posts. Then `GET /blogs/1?include=posts&fields[blogs]=posts&page[number]=posts:2` returns the next 10 posts.

One reason to model large relationships is that it enables to query them. I consider exposing an `eventId` attribute an anti-pattern, it's like modeling foreign keys instead of using JSON:API relationships, so the client needs to know dependencies that aren't exposed through HATEOS links. In your example, the client needs to "infer" that `article.eventId` maps to `event.id`. As a human, this may seem obvious, but the relationship isn't defined for automation scenarios. So a generic API explorer wouldn't be able to offer clicking through the data, which is what HATEOS is all about. Beware that the API developer and API consumer may not even know each other.

Using our same approach for responding to updates would not be very useful, as it potentially only returns the first page of relationship data. But the hard part is that _detecting_ changes requires analyzing the thousands of related records. I find it odd to _require_ returning data for relationships the client may never care about. For attribute data, this is not a concern because the set of members is small and fetching them isn't expensive.

Caching is all about storing a _small subset_ of data you access frequently. It is never about keeping a full database snapshot around, which is what the spec seems to enforce when it comes to relationships.

**Comment by jelhan** [comment]
Do I understand correctly that you are paginating the collection of related resource objects returned for a `GET` request to a relationship URL? As far as I understand the specification, pagination is also allowed for the response to an update of a relationship request through a relationship URL.

> hard part is that _detecting_ changes requires analyzing the thousands of related records.

The specification does not require that you detect it. You can always response with `200 Ok` and response with updated resource (on updates through resource URL) or relationship (on updates through relationship URL). Only if responding with a `204 No Content` you must ensure that there wasn't another change to resource / relationship than those requested explicitly by the client.

> It is never about keeping a full database snapshot around, which is what the spec seems to enforce when it comes to relationships.

I cannot follow this one. If I got it right, you are paginating the relationship data when responding to a `GET` request to a relationship URL. In my reading of the specification this is allowed. Even though wording is not very precise regarding pagination of relationship URLs.

**Comment by bkoelman** [comment]
> Do I understand correctly that you are paginating the collection of related resource objects returned for a GET request to a relationship URL?

No, we are paginating the type/id pairs in the data array (the spec calls them relationship elements). A GET request to a relationship URL doesn't return resources.

> As far as I understand the specification, pagination is also allowed for the response to an update of a relationship request through a relationship URL.

Yes, but doing so is pointless as explained earlier, so we're not taking the associated performance hit. SQL joins are expensive, so unless the user explicitly asks for them, we avoid them. A resource may have 10 relationships, but a GET request typically includes only one or two. Now we need to fetch all 10 for no reason. Pagination just makes it less of a problem (return 500 records when page size = 50), yet it's still unnecessary extra costs in terms of server processing and network bandwidth.

> The specification does not require that you detect it. You can always response with 200 Ok and response with updated resource (on updates through resource URL) or relationship (on updates through relationship URL). Only if responding with a 204 No Content you must ensure that there wasn't another change to resource / relationship than those requested explicitly by the client.

So it's choosing between expensive detection -or- sending unneeded data for *all* relationships to the client. Neither sounds attractive.

> I cannot follow this one.

It is not a useful response for a client when the server returns the first 10 relationship elements from a set of thousands, meaning that "something" in those thousands has changed. The change may be in the first 10 or not, there's no way for the client to know, unless it fetches them all.

Client caching is usually about a small subset, for example only what applies to the current user. So caching the full relationship isn't the practice; instead filters are used to only fetch what's needed for the current user. Therefore returning all elements of all relationships (even when paginated) does not fit typical use cases.

**Comment by lode** [comment]
To me this is the main argument:

> (...) I don't see how that's useful to a client: it won't enable cache refresh, because it doesn't include side-effect attachment/detachment of related resources that weren't in the request. For example: if a user removes one phone number from a contact, it won't observe that a concurrent user added two new phone numbers to that same contact.

Because it is not about the spec, but about what's interesting to implement.

The requirement for the server to send the resource/relationship in the response on creating/updating resources/relationships is maybe only logical for creating resources. Then the client had some data, the server modified it, and no-one else could have been in between. So it is good if the server responds with the modified resource to allow the client to continue to work with that data.

For updating resources it already becomes a little ambiguous. Returning only the modifications done during the request doesn't guarantee any modifications done by other clients in between. So what's the point? Not that it should be forbidden, there might be use cases where it makes sense. But forcing it sort of suggests that the client can continue with the data / has the most fresh data, which isn't in the spec right now, and also can cause more issues than it solves (when data from other clients is returned that need more context to deal with properly).

The same counts for creating/updating relationships. Other clients could have changed things and thus returning the modifications from the request might not be very useful. And here it is even more important than when updating resources, since as mentioned relationships can get rather expensive.

Maybe it is a good idea to change all MUSTs related to responding with modified resources/relationships to MAYs? Or maybe to SHOULDs since there's probably more use case where you can benefit from the modified data than not? Only for creating resources it might be worthwhile to keep it as a MUST.

And then allow the client to use GETs if they want to get fresh information.

---

Just a small remark, you mentioned `Not Modified`:

> Therefore, I'd like the MUST in 200 relationship responses to be changed to MAY, so that our current behavior is not a violation. In other words: allow to return Not Modified, even if there are potential side effects. Right now the spec guarantees clients there are no side effects when Not Modified is returned, which I find unrealistic to require that guarantee from servers.

Didn't you mean `No Content`?

As far as I understand `Not Modified` is for case when the client sends a request, e.g. to add a relationship, and the server already had that relationship. Thus it replies with `Not Modified`. If the server didn't already have the relationship it should reply with `OK` or `No Content`, depending on whether other attributes changed.

**Comment by bkoelman** [comment]
Yes, I meant 204 No Content.

**Comment by bkoelman** [comment]
@jelhan I've gone to great lengths trying to motivate my point of view and answer your follow-up questions. Yet this issue is still closed as completed. I've stated possible interpretations in https://github.com/json-api/json-api/issues/1524#issuecomment-1847063903, but didn't get a response on how it is supposed to be interpreted. Wouldn't that be reason enough to reopen this issue?

I also demonstrated why keeping caches up-to-date doesn't make sense in practice _when it comes to relationships_, and requiring this from servers is unrealistic, asking to change MUST to MAY. I haven't seen a motivation that proves me wrong or a legitimate use case I wasn't aware of. But also not an acknowledgement from you or any of the spec authors that this part of the spec should be reconsidered.

Were all my attempts here a wasted effort?
